src/NuGet.Clients/NuGet.SolutionRestoreManager/ErrorListTableDataSource.cs

```diff
@@ -0,0 +1,208 @@
+﻿using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Shell.TableControl;
+using Microsoft.VisualStudio.Shell.TableManager;
+using NuGet.PackageManagement.UI;
+
+namespace NuGet.SolutionRestoreManager
+{
+    /// <summary>
+    /// Add/Remove warnings/errors from the error list.
+    /// This persists messages once they are added.
+    /// </summary>
+    [Export(typeof(ErrorListTableDataSource))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    internal sealed class ErrorListTableDataSource : ITableDataSource, IDisposable
+    {
+        private readonly object _initLockObj = new object();
+        private readonly object _subscribeLockObj = new object();
+        private readonly IServiceProvider _serviceProvider;
+        private IReadOnlyList<TableSubscription> _subscriptions = new List<TableSubscription>();
+
+        public string SourceTypeIdentifier => StandardTableDataSources.ErrorTableDataSource;
+
+        public string Identifier => "NuGetRestoreManagerListTable";
+
+        public string DisplayName => "NuGet_Restore_Manager_Table_Data_Source";
+
+        private IErrorList _errorList;
+        private ITableManager _tableManager;
+        private bool _initialized;
+
+        [ImportingConstructor]
+        public ErrorListTableDataSource(
+            [Import(typeof(SVsServiceProvider))]
+            IServiceProvider serviceProvider)
+        {
+            if (serviceProvider == null)
+            {
+                throw new ArgumentNullException(nameof(serviceProvider));
+            }
+
+            _serviceProvider = serviceProvider;
+        }
+
+        public IDisposable Subscribe(ITableDataSink sink)
+        {
+            var subscription = new TableSubscription(sink);
+
+            lock (_subscribeLockObj)
+            {
+                // Add valid subscriptions to the new list to start
+                var updatedSubscriptions = new List<TableSubscription>(_subscriptions.Where(e => !e.Disposed))
+                {
+                    // Add the new subscription
+                    subscription
+                };
+
+                // Apply changes
+                _subscriptions = updatedSubscriptions;
+            }
+
+            return subscription;
+        }
+
+        /// <summary>
+        /// Clear only nuget entries.
+        /// </summary>
+        public void ClearNuGetEntries()
```

Comment:
Would be better with consistent naming:
AddNuGetEntries/ClearNuGetEntries or AddEntries/ClearEntries